jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpChannelOverHTTP2.java

```diff
@@ -261,31 +275,277 @@ public InvocationType getInvocationType()
             {
                 return callback.getInvocationType();
             }
-        });
+        };
+        boolean wakeup = _contentDemander.onContent(content);
+        boolean handle = onContent(content);
 
         boolean endStream = frame.isEndStream();
         if (endStream)
         {
             boolean handleContent = onContentComplete();
+            // this will generate EOF -> must happen before onContentProducible
             boolean handleRequest = onRequestComplete();
             handle |= handleContent | handleRequest;
         }
 
+        boolean woken = wakeup && getRequest().getHttpInput().onContentProducible();
         if (LOG.isDebugEnabled())
         {
-            LOG.debug("HTTP2 Request #{}/{}: {} bytes of {} content, handle: {}",
+            LOG.debug("HTTP2 Request #{}/{}: {} bytes of {} content, woken: {}, handle: {}",
                     stream.getId(),
                     Integer.toHexString(stream.getSession().hashCode()),
                     length,
                     endStream ? "last" : "some",
+                    woken,
                     handle);
         }
+        handle |= woken;
 
         boolean wasDelayed = _delayedUntilContent;
         _delayedUntilContent = false;
         return handle || wasDelayed ? this : null;
     }
 
+    private static final HttpInput.Content EOF = new HttpInput.EofContent();
+
+    private static final class DemandingContent extends HttpInput.SpecialContent
+    {
+        private final boolean wakeup;
+
+        public DemandingContent(boolean wakeup)
+        {
+            this.wakeup = wakeup;
+        }
+    }
+
+    private class ContentDemander
+    {
+        private final AtomicReference<HttpInput.Content> _content = new AtomicReference<>();
+
+        public void recycle()
+        {
+            HttpInput.Content c = _content.getAndSet(null);
+            if (c != null && !c.isSpecial())
+                throw new AssertionError("unconsumed content: " + c);
+        }
+
+        public HttpInput.Content poll()
+        {
+            while (true)
+            {
+                HttpInput.Content c = _content.get();
+                if (c == null || c.isSpecial() || _content.compareAndSet(c, c.isEof() ? EOF : null))
+                    return c;
+            }
+        }
+
+        public boolean demand(boolean wakeup)
+        {
+            while (true)
+            {
+                HttpInput.Content c = _content.get();
+                if (c instanceof DemandingContent)
+                {
+                    if (wakeup && !((DemandingContent)c).wakeup)
+                    {
+                        if (!_content.compareAndSet(c, new DemandingContent(true)))
+                            continue;
+                    }
+                    return false;
+                }
+                if (c != null)
+                    return true;
+                if (_content.compareAndSet(null, new DemandingContent(wakeup)))
```

Comment:
Create 2 static final DemandingContents and pick between them rather than creating new ones all the time

Reply:
Done.